Created my own ArrayList class, which implements the List interface. …

MaritaPasha opened the pull request:
…Override the methods required. Add a method expandCapacity, which extends the array

MaritaPasha committed:
Created my own ArrayList class, which implements the List interface. Override the methods required. Add a method expandCapacity, which extends the array

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,101 @@
 package core.basesyntax;
 
```
This violates checklist item #1: "Do not begin class or method implementation with an empty line". Remove the empty line after the package declaration.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,101 @@
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+
```
This violates checklist item #1: "Do not begin class or method implementation with an empty line". Remove the empty line after the class declaration.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,101 @@
+
+    private void expandCapacity() {
+        if (size == elementsData.length) {
+            int newCapacity = (int) (elementsData.length * 1.5);
```
This violates checklist item #3: "Replace magic numbers with named constants". The growth factor 1.5 should be a named constant, not a magic number.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,101 @@
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > size) {
```
This violates checklist item #5: "Move repeated index checks to a private method". The index validation logic is repeated in multiple methods. Extract private helper methods for access and insertion index validation.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,101 @@
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0 || index >= size) {
```
This violates checklist item #5: "Move repeated index checks to a private method". The index validation logic is repeated in multiple methods. Extract private helper methods for access and insertion index validation.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,101 @@
     @Override
     public void set(T value, int index) {
-
+        if (index < 0 || index >= size) {
```
This violates checklist item #5: "Move repeated index checks to a private method". The index validation logic is repeated in multiple methods. Extract private helper methods for access and insertion index validation.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,101 @@
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0 || index >= size) {
```
This violates checklist item #5: "Move repeated index checks to a private method". The index validation logic is repeated in multiple methods. Extract private helper methods for access and insertion index validation.

mateacademy-ai-mentor reviewed (changes requested):
### Review Summary

**Decision:** REQUEST_CHANGES — there are checklist violations that are considered core requirements for this task.

**Why:** Your implementation is functionally solid, but it violates explicit checklist items, which are required for approval. Specifically: magic numbers, repeated index checks, and style rules about leading empty lines.

**Critical issues to fix (blockers):**
1. Checklist #3 – Replace magic numbers with named constants
   - Location: `ArrayList.java`, in `expandCapacity` (e.g., around the line where new capacity is computed using `1.5`).
   - Why: Using `1.5` directly reduces clarity and violates the checklist.
   - How: Introduce a private constant, for example `private static final double GROWTH_FACTOR = 1.5;`, and use it in capacity calculations.
2. Checklist #5 – Move repeated index checks to private helper methods
   - Locations: `get(int index)`, `set(int index, T value)`, `remove(int index)`, and `add(int index, T value)` each repeat index validation logic (look for repeated `if (index < 0 || index >= size)` and `if (index < 0 || index > size)` patterns).
   - Why: Repetition increases maintenance risk and violates the checklist.
   - How: Extract two private methods, e.g., `checkElementIndex(int index)` for access/update/remove (valid range: 0..size-1) and `checkPositionIndex(int index)` for insertion (valid range: 0..size). Replace inline checks with these helpers and keep consistent exception messages.
3. Checklist #1 – Do not begin class or method implementation with an empty line
   - Locations: `ArrayList.java` after the `package` declaration, after the `class` declaration, and at the start of some method bodies.
   - Why: Violates required style rules.
   - How: Remove leading blank lines at the start of the class and any method implementations.

**Minor improvements (non-blockers, but recommended):**
- Rename `elementsData` to `elementData` to match common Java naming in array-backed lists.

You’ve done well with core functionality, resizing, and exception messages. Address the above items and you should be all set. Keep going—this is very close to approval!

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

> [!IMPORTANT]
> Fix the issues, commit the changes, and then re-request my review using the **`Re-request review` button**[^2] in the Reviewers tab.

[^2]: [Re-request AI review example](https://snipboard.io/nOXxHN.jpg)

> [!TIP]
> If you have any comments about this review, please leave a message in the [Add a comment](#new_comment_field) section below and **mention me with the `@mateacademy-ai-mentor` tag**[^3].
> I will read your message during my next review of your pull request. Please do not reply directly to my comments, as that feature is not supported yet.

[^3]: [Add AI review comments example](https://snipboard.io/hVlOey.jpg)

> [!CAUTION]
> Do not close the pull request until the review is complete. Closing it early can lead to unexpected behavior and disrupt the review process.

MaritaPasha committed:
Implemented the requested changes:

Replaced magic numbers with named constants – introduced private static final double GROWTH_FACTOR = 1.5 and used it in capacity calculations.

Extracted repeated index checks into private helper methods – added checkElementIndex(int index) for access/update/remove and checkPositionIndex(int index) for insertions. All relevant methods now use these helpers.

Removed leading empty lines – cleaned up class and method declarations to comply with style rules.

Renamed elementsData to elementData – aligned with common Java naming conventions for array-backed lists.

mateacademy-ai-mentor reviewed (approved):
### Review Summary

**Decision: APPROVE**

Your custom `ArrayList` implementation fully meets the functional and style requirements. You’ve addressed prior issues by introducing `DEFAULT_CAPACITY` and `GROWTH_FACTOR`, centralizing index validation (`checkPositionIndex`, `checkElementIndex`), and using `System.arraycopy` for shifts. Exception handling is correct and informative, including `ArrayListIndexOutOfBoundsException` for invalid indices and `NoSuchElementException` in `remove(T element)` when not found.

Highlights:
- Clear, maintainable code with no magic numbers or redundant logic.
- Correct 1.5x growth behavior and accurate size tracking.
- All edge cases (empty list operations, boundary inserts/removes) behave as expected.

No critical blockers or missing requirements were found. Nice work—this is clean, robust, and ready to move forward!

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)
